pkg/gpu/consumer.go

```diff
@@ -0,0 +1,168 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	gpuebpf "github.com/DataDog/datadog-agent/pkg/gpu/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/process/monitor"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// cudaEventConsumer is responsible for consuming CUDA events from the eBPF probe, and delivering them
+// to the appropriate stream handler.
+type cudaEventConsumer struct {
+	eventHandler                    ddebpf.EventHandler
+	once                            sync.Once
+	closed                          chan struct{}
+	streamHandlers                  map[model.StreamKey]*StreamHandler
+	wg                              sync.WaitGroup
+	scanTerminatedProcessesInterval time.Duration
+}
+
+// NewCudaEventConsumer creates a new CUDA event consumer.
+func NewCudaEventConsumer(eventHandler ddebpf.EventHandler, scanTerminatedProcessesInterval time.Duration) *cudaEventConsumer {
+	return &cudaEventConsumer{
+		eventHandler:                    eventHandler,
+		closed:                          make(chan struct{}),
+		streamHandlers:                  make(map[model.StreamKey]*StreamHandler),
+		scanTerminatedProcessesInterval: scanTerminatedProcessesInterval,
+	}
+}
+
+// Stop stops the CUDA event consumer.
+func (c *cudaEventConsumer) Stop() {
+	if c == nil {
+		return
+	}
+	c.eventHandler.Stop()
+	c.wg.Wait()
+	c.once.Do(func() {
+		close(c.closed)
+	})
```

Comment:
Could calling wait before closing the channel create a deadlock? I see we call `c.wg.Done()` inside `Start()` and to exit the start loop is by closing the `c.closed` channel

Reply:
Should we add unit test for `NewCudaEventConsumer` to warranty that things are working as expected? Preferably adding `-race` flag 😄

Reply:
The order of the operations was wrong, I added a test to check that we can start/stop the consumer without issues, nice catch :D